dll/win32/shlwapi/appcompat.c

```diff
@@ -0,0 +1,467 @@
+    { "ue32.exe", "2.00.0.0", SHACF_OLDREGITEMGDN },
+    { "wpwin7.exe", NULL, SHACF_CONTEXTMENU | SHACF_CORELINTERNETENUM },
+    { "wpwin8.exe", NULL, SHACF_CORELINTERNETENUM | SHACF_OLDREGITEMGDN },
+    { "wpwin9.exe", MAJOR_VER_ONLY "9", SHACF_CORELINTERNETENUM },
+};
+
+// Window class name and compatibility flags
+typedef struct tagWNDCOMPATINFO
+{
+    PCSTR pszLengthAndClassName;
+    DWORD dwCompatFlags;
+} WNDCOMPATINFO, *PWNDCOMPATINFO;
+static WNDCOMPATINFO g_wndCompatInfo[] =
+{
+    // The first byte indicates the string length
+    { "\x09" "bosa_sdm_", SHACF_UNKNOWN1 | SHACF_UNKNOWN2 },
+    { "\x18" "File Open Message Window", SHACF_UNKNOWN1 | SHACF_UNKNOWN2 },
+};
+
+// Internal structure for SHLWAPI_WndCompatEnumProc
+typedef struct tagAPPCOMPATENUM
+{
+    PWNDCOMPATINFO pItems;
+    SIZE_T nItems;
+    DWORD dwProcessId;
+    INT iFound;
+} APPCOMPATENUM, *PAPPCOMPATENUM;
+
+static BOOL CALLBACK
+SHLWAPI_WndCompatEnumProc(_In_ HWND hWnd, _In_ LPARAM lParam)
+{
+    PAPPCOMPATENUM pEnum = (PAPPCOMPATENUM)lParam;
+
+    CHAR szClass[256];
+    if (!pEnum->nItems || !GetClassNameA(hWnd, szClass, _countof(szClass)))
+        return TRUE; // Ignore and continue
+
+    // Search for the target window in pEnum
+    const INT cchClass = lstrlenA(szClass);
+    for (UINT iItem = 0; iItem < pEnum->nItems; ++iItem)
+    {
+        PCSTR pszLengthAndClassName = pEnum->pItems[iItem].pszLengthAndClassName;
+
+        INT cchLength = pszLengthAndClassName[0]; // The first byte represents the length
+        if (cchClass < cchLength)
+            cchLength = cchClass; // Ignore trailing characters
+
+        // Compare the string
+        if (StrCmpNA(szClass, &pszLengthAndClassName[1], cchLength) == 0) // Class name matched
+        {
+            // Get the process ID
+            DWORD dwProcessId;
+            GetWindowThreadProcessId(hWnd, &dwProcessId);
+            if (dwProcessId == pEnum->dwProcessId) // Same process
+            {
+                pEnum->iFound = iItem; // Found
+                return FALSE; // Quit enumeration
+            }
+        }
+    }
+
+    return TRUE; // Continue enumeration
+}
+
+// English (US) UTF-16
+#define PRODUCT_VER_ENGLISH_US_UTF16    "\\StringFileInfo\\040904E4\\ProductVersion"
+// German (Germany) UTF-16
+#define PRODUCT_VER_GERMAN_UTF16        "\\StringFileInfo\\040704E4\\ProductVersion"
+// English (US) Western European
+#define PRODUCT_VER_ENGLISH_US_WE       "\\StringFileInfo\\040904B0\\ProductVersion"
+// English (US) Neutral
+#define PRODUCT_VER_ENGLISH_US_NEUTRAL  "\\StringFileInfo\\04090000\\ProductVersion"
+// Swedish (Sweden) Western European
+#define PRODUCT_VER_SWEDISH_WE          "\\StringFileInfo\\041D04B0\\ProductVersion"
+
+typedef struct tagLANGANDCODEPAGE
+{
+    WORD wLanguage;
+    WORD wCodePage;
+} LANGANDCODEPAGE, *PLANGANDCODEPAGE;
+
+static HRESULT
+SHLWAPI_GetModuleVersion(_In_ PCSTR pszFileName, _Out_ PSTR *ppszDest)
+{
+    DWORD dwHandle;
+    BYTE Data[4096];
```

Comment:
4096? That's a bit _too_ large, you could use MAX_PATH or something, it's smaller, and if in the rare case you would need a larger buffer, you could just allocate a dynamic one.
See for example: https://git.reactos.org/?p=reactos.git;a=blob;f=base/applications/msconfig_new/utils.c;hb=b5f763447c00049b954b7a77896813a8f6f2d128#l305